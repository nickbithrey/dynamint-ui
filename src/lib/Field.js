import React from 'react';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';

class Field extends React.Component {
	
	constructor(props) {
		super(props);
		this.state = {key: props.name, value: props.value};
		this.updateValue = this.updateValue.bind(this);
		this.update = this.update.bind(this);
	}
	
	updateValue(e) {
		e.preventDefault();
		this.setState({...this.state, value: e.target.value});
	}
	
	update(e) {
		this.props.update({[this.state.key]: this.state.value}, e);
	}
	
	render() {
		if (this.props.checkbox) {
			return <Checkbox {...this.props} defaultChecked={this.state.value} onChange={this.updateValue} onBlur={this.update} />
		}
		return (
			<TextField {...this.props} value={this.state.value} onChange={this.updateValue} onBlur={this.update} />
		);
	}
	
}

export default Field;