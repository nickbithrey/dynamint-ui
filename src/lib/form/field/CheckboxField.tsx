import * as React from 'react';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';
import { updateOnChange, ITypedField } from './Field';

export const CheckboxField = ({name, label, value, update}: ITypedField<boolean>) => (
        <Checkbox 
            label={label}
            defaultChecked={value}
            onChange={updateOnChange(name, update)}
        />
);

