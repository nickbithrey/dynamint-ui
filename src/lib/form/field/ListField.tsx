import * as React from 'react';
import { updateOnChange, ITypedField } from './Field';
import { IButton, ButtonContainer } from 'lib/button';
import { Key, Condition } from '~/lib';
import Table, { ITableColumn } from 'lib/table';
import { Selector, ISelectionDetails, ISelectionButton, SelectionType } from 'lib/selector';

export interface IListField extends ITypedField<Array<Object>> {
    storeKey: string;
    loading: boolean;
    entityForm: React.ComponentType;
    columns: Array<ITableColumn>;
}

const createButton: ISelectionButton & Condition<any> = {
    button: {
        key: 'create',
        content: 'Create',
        link: '/compconfigs',
        type: 'default'
    },
    condition: (input: ISelectionDetails) => input.selected.length === 0
};

interface IListFieldState {
    items: Array<any>;
    selectedItem?: any;
    selectedIndex?: number;
}

export class ListField extends React.Component<ITypedField<Array<Object>>, IListFieldState> {

    constructor(props: ITypedField<Array<Object>>) {
        super(props);
        this.state = {items: props.value};
        this.buttons = this.buttons.bind(this);
        this.onEditDetails = this.onEditDetails.bind(this);
        this.renderForm = this.renderForm.bind(this);
        this.updateField = this.updateField.bind(this)
        this.clearEditDetails = this.clearEditDetails.bind(this);
        this.createNewItem = this.createNewItem.bind(this);
    }
    
    buttons(): Array<ISelectionButton & Condition<any>> {
        return [
            {
                button: {
                    key: 'edit', 
                    content: 'Edit',
                    type: 'primary',
                },
                onClickResolver: this.onEditDetails,
                condition: (input: ISelectionDetails) => input.selected.length > 0,
             },
             {
                 button: {
                     key: 'create',
                     content: 'Create',
                     type: 'default',
                     onClick: this.createNewItem
                 },
                 condition: (input: ISelectionDetails) => input.selected.length === 0
             }
        ];
    }
    
    createNewItem(): void {
        this.onEditDetails({
            selectedCount: 1,
            selected: [{}],
            selectedIndices: [this.state.items.length]
        })();
    }
    
    updateField(item: any): void {
        const list = [...this.state.items];
        list[this.state.selectedIndex] = item;
        this.props.update(this.props.name, list);
        this.clearEditDetails();
        this.setState({
            items: list
        });
    }
    
    onEditDetails(input: ISelectionDetails): () => void {
        return () => {
            this.setState({selectedItem: input.selected[0], selectedIndex: input.selectedIndices[0]});
        }
    }
    
    clearEditDetails(): void {
        this.setState({
            selectedItem: undefined,
            selectedIndex: undefined
        });
    } 
    
    render() {
        const {
            loading,
            columns,
            entityForm,
            storeKey
        } = this.props;
        const {
            items
        } = this.state;
        return (
            <>
                <Selector
                    storeKey={storeKey}
                    loading={loading}
                    items={items}
                    columns={columns}
                    buttons={this.state.selectedItem ? [] : this.buttons()}
                    selectionType={SelectionType.single}
                />
                {
                    this.renderForm(entityForm)
                }
            </>
        );
    }
    
    renderForm(entityForm: React.ComponentType<any>) {
        const Form = entityForm;
        if (this.state.selectedItem) {
            return <Form 
                item={this.state.selectedItem}
                update={this.updateField}
                back={this.clearEditDetails}
            />;
        }
        return null;
    }
    
}

