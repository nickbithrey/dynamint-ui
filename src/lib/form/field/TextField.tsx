import * as React from 'react';
import { TextField as FTextField } from 'office-ui-fabric-react/lib/TextField';
import { updateOnChange, ITypedField } from './Field';

export const TextField = ({name, label, value, update}: ITypedField<string>) => (
        <FTextField 
            label={label}
            defaultValue={value}
            onChange={updateOnChange(name, update)}
        />
);

