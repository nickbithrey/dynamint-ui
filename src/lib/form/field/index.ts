import { Field, IField, IFieldType, IFieldValueType, ITypedField } from './Field';

export {
    Field,
    IField,
    IFieldType, 
    IFieldValueType, 
    ITypedField
};