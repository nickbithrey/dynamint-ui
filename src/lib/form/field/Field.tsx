import * as React from 'react';
import { TextField } from './TextField';
import { CheckboxField } from './CheckboxField';
import { ListField } from './ListField';
import { IURILink } from '~/lib';

export const updateOnChange = <V extends IFieldValueType>(name: string, update: (name: string, value: V) => void) => (
    (e: React.ChangeEvent<HTMLInputElement>) => {
        e.preventDefault();
        update(name, e.target.value as unknown as V)
    }
);

export type IFieldType = 'text' | 'checkbox' | 'list';

export type IFieldValueType = string | number | boolean | Array<any> | IURILink;

export interface IField<V extends IFieldValueType> extends ITypedField<V> {
    type: IFieldType;
}

export interface ITypedField<V extends IFieldValueType> {
    name: string;
    label?: string;
    value: V;
    update?: (key: string, value: V) => void;
    entity?: any;
    [key: string]: any;
}

const fieldTypes: { [key in IFieldType]: React.ComponentType<any>} = {
    text: TextField,
    checkbox: CheckboxField,
    list: ListField,
};

export const Field = <V extends IFieldValueType>({type, ...remainingProps}: IField<V>) => {
    const FieldType: React.ComponentType<ITypedField<IFieldValueType>> = fieldTypes[type];
    return <FieldType {...remainingProps} />
}