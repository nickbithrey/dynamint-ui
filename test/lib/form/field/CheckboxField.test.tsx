import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ITypedField } from 'lib/form/field/Field';
import { CheckboxField } from 'lib/form/field/CheckboxField';

describe('<CompConfigAttrDetails />', () => {
    
    const props: ITypedField<boolean> = {
        name: 'name',
        label: 'label',
        value: true,
        update: jest.fn(),
        entity: {},
    };

    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = shallow(<CheckboxField {...props} />);
    });
    
    it('renders a valid tag and matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
});