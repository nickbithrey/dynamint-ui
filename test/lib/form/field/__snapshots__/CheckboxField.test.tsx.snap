// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CompConfigAttrDetails /> renders a valid tag and matches snapshot 1`] = `
<StyledCheckboxBase
  defaultChecked={true}
  label="label"
  onChange={[Function]}
/>
`;
